src/Servers/IIS/IISIntegration/src/IISMiddleware.cs

```diff
@@ -147,5 +156,40 @@ public async Task Invoke(HttpContext httpContext)
 
             await _next(httpContext);
         }
+
+        private WindowsPrincipal GetUser(HttpContext context)
+        {
+            var tokenHeader = context.Request.Headers[MSAspNetCoreWinAuthToken];
+
+            if (!StringValues.IsNullOrEmpty(tokenHeader)
+                && int.TryParse(tokenHeader, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var hexHandle))
+            {
+                // Always create the identity if the handle exists, we need to dispose it so it does not leak.
+                var handle = new IntPtr(hexHandle);
+                var winIdentity = new WindowsIdentity(handle, IISDefaults.AuthenticationScheme);
+
+                // WindowsIdentity just duplicated the handle so we need to close the original.
+                NativeMethods.CloseHandle(handle);
+
+                context.Response.OnCompleted(ClearUserDelegate, (context, winIdentity));
```

Comment:
FYI: This allocates now. Can you shove that state into the HttpContext.Items?

Reply:
Or double register
```csharp
context.Response.OnCompleted(DisposeUserDelegate, winIdentity);
context.Response.OnCompleted(ClearUserDelegate, context);
```

Reply:
I thought about that but didn't suggest because I was thinking it's another allocation. I guess it isn't because the list default size is 4.

Reply:
I assume the `.User` on HttpContext can't be trusted? (as it may have changed and the one that needs disposing lost?)